src/js/quick-hull-3d/face.js

```diff
@@ -0,0 +1,271 @@
+                lenSqrMax = lenSqr;
+            }
+
+            halfEdge = halfEdge.next;
+        } while (halfEdge !== this.halfEdge0);
+
+        headPoint = halfEdgeMax.head().point;
+        tailPoint = halfEdgeMax.tail().point;
+
+        lenMax = Math.sqrt(lenSqrMax);
+
+        u = List.scaldiv(CSNumber.real(lenMax), List.sub(headPoint, tailPoint));
+        dot = List.scalproduct(u, this.normal);
+
+        this.normal = List.sub(this.normal, List.scalmul(dot, u));
+    }
+};
+
+Face.prototype.getEdge = function(index) {
+    var halfEdge = this.halfEdge0;
+
+    while (index > 0) {
+        index--;
+        halfEdge = halfEdge.next;
+    }
+
+    while (index < 0) {
+        index++;
+        halfEdge = halfEdge.previous;
+    }
+
+    return halfEdge;
+};
+
+Face.prototype.getFirstEdge = function() {
+    return this.halfEdge0;
+};
+
+Face.prototype.findEdge = function(tailVertex, headVertex) {
+    var halfEdge = this.halfEdge0;
+
+    do {
+        if (halfEdge.tail() === tailVertex && halfEdge.head() === headVertex) {
+            return halfEdge;
+        }
+
+        halfEdge = halfEdge.next;
+    } while (halfEdge !== this.halfEdge);
+
+};
+
+Face.prototype.distanceToPlane = function(point) {
+    List.scalproduct(this.normal, point);
+};
+
+Face.prototype.getVertexString = function() {
+    var result = this.halfEdge0.head().index,
+        halfEdge = this.halfEdge0.next;
+
+    while (halfEdge !== this.halfEdge0) {
+        result += ' ' + halfEdge.head().index;
+        halfEdge = halfEdge.next;
+    }
+
+    return result;
+};
+
+Face.prototype.getVertexIndices = function() {
+    var result = [],
+        halfEdge = this.halfEdge0;
+
+    do {
+        result.push(halfEdge.head().index);
+        halfEdge = halfEdge.next;
+    } while (halfEdge !== this.halfEdge0);
+
+    return result;
+};
+
+Face.prototype.checkConsistency = function() {};
+
+Face.prototype.mergeAdjacentFace = function(adjacentEdge, discardedFaces) {
+    var numberOfDiscarded = 0,
+        oppositeFace = adjacentEdge.oppositeFace();
+
+    discardedFaces[numberOfDiscarded++] = oppositeFace;
+
+};
+
+Face.prototype.getSquaredArea = function(halfEdge0, halfEdge1) {};
+Face.prototype.triangulate = function(newFaces, minArea) {};
+
+Face.prototype._computeNormalAndCentroid = function(minArea) {
+    var numberOfVertices, halfEdge;
+
+    this.computeNormal(minArea);
+    this.computeCentroid();
+    this.planeOffset = List.scalproduct(this.normal, this.centroid);
+
+    if (minArea !== void 0) {
+        numberOfVertices = 0;
+        halfEdge = this.halfEdge0;
+
+        do {
+            numberOfVertices++;
+            halfEdge = halfEdge.next;
+        } while (halfEdge !== this.halfEdge0);
+
+        if (numberOfVertices !== this.numberOfVertices) {
+            throw new Error('Face ' + this.getVertexString() + ' should be ' + this.numberOfVertices);
+        }
+    }
+};
+
+Face.prototype._connectHalfEdges = function(hedgePrev, hedge) {
```

Comment:
I do not use this function right now. 
<a href="https://github.com/CindyJS/CindyJS/blob/master/src/java/cindyjs/quickhull3d/Face.java#L330">java version</a>